File: Blind75-Practice/easy/03-stack/Python/implement-stack-using-queues.py
"""
LeetCode 225: Implement Stack using Queues
Difficulty: Easy
Pattern: Stack

Implement LIFO stack using only queue operations (FIFO).

Approach: One deque — push appends; pop/peek rotate size-1 times so tail becomes head (amortized
cost per op can be O(n) for naive single-queue rotation, or use two queues / lazy top).

This version: collections.deque for O(1) popleft/append; push O(1); pop O(n) by rotating.

Time Complexity: push O(1), pop/top O(n) worst case per call (acceptable for practice)
Space Complexity: O(n)
"""

from collections import deque


class MyStack:
    def __init__(self) -> None:
        self._q: deque[int] = deque()

    def push(self, x: int) -> None:
        self._q.append(x)

    def pop(self) -> int:
        for _ in range(len(self._q) - 1):
            self._q.append(self._q.popleft())
        return self._q.popleft()

    def top(self) -> int:
        for _ in range(len(self._q) - 1):
            self._q.append(self._q.popleft())
        x = self._q[0]
        self._q.append(self._q.popleft())
        return x

    def empty(self) -> bool:
        return len(self._q) == 0


if __name__ == "__main__":
    st = MyStack()
    st.push(1)
    st.push(2)
    assert st.top() == 2
    assert st.pop() == 2
    assert st.empty() is False
    assert st.pop() == 1
    assert st.empty() is True
    print("All test cases passed!")

File: Blind75-Practice/easy/03-stack/Java/ImplementStackUsingQueues.java
/*
 * 225. Implement Stack using Queues
 * Difficulty: Easy
 * Pattern: Stack / Queue Simulation
 *
 * Problem
 * -------
 * Implement a last-in-first-out (LIFO) stack using only queue operations. The stack
 * should support the normal operations:
 *
 * - `push(x)`: push `x` onto the top of the stack
 * - `pop()`: remove and return the top element
 * - `top()`: return the top element without removing it
 * - `empty()`: return whether the stack is empty
 *
 * You may use only standard queue operations such as push to back, peek/pop from front,
 * size, and checking whether the queue is empty.
 *
 * Example 1
 * ---------
 * Input:
 * ["MyStack", "push", "push", "top", "pop", "empty"]
 * [[],        [1],    [2],    [],    [],    []]
 * Output:
 * [null, null, null, 2, 2, false]
 *
 * Constraints
 * -----------
 * - 1 <= x <= 9
 * - At most 100 calls will be made to `push`, `pop`, `top`, and `empty`
 * - All calls to `pop` and `top` are valid
 *
 * How to think about it
 * ---------------------
 * A stack needs access to the most recently added value, but a queue naturally exposes
 * the oldest value first. One clean simulation is to keep everything in one queue and,
 * whenever you need the top, rotate the first `n - 1` items to the back so the newest
 * item moves to the front. That makes `pop()` and `top()` more expensive, but keeps the
 * behavior correct using queue-only operations.
 *
 * Time: push O(1), pop/top O(n), Space: O(n)
 */

import java.util.ArrayDeque;
import java.util.Deque;

public class ImplementStackUsingQueues {

    static class MyStack {
        private final Deque<Integer> q = new ArrayDeque<>();

        public void push(int x) {
            q.push(x);
            for(int i = 0; i < q.size() -1; i++) {
                q.push(q.removeFirst());
            }
            // 

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            // q.addLast(x);
        }

        public int pop() {
            return q.remove();
            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            // for (int i = q.size() - 1; i > 0; i--) {
                // q.addLast(q.removeFirst());
            // }
            // return q.removeFirst();
        }

        public int top() {
            return q.peek();
            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            // [1 2 3 4]

            // [2 3 4 1]
            // [3 4 1 2]
            // [4 1 2 3]

            // for (int i = q.size() - 1; i > 0; i--) {
            //     q.addLast(q.removeFirst());
            // }
            // int value = q.peekFirst();
            // q.addLast(q.removeFirst());
            // return value;
        }

        public boolean empty() {
            return q.isEmpty();
            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            // return q.isEmpty();
        }
    }

    public static void main(String[] args) {
        MyStack st = new MyStack();
        st.push(1);
        st.push(2);
        assert st.top() == 2;
        assert st.pop() == 2;
        assert !st.empty();
        assert st.pop() == 1;
        assert st.empty();
        System.out.println("All test cases passed!");
    }
}
